pvlib/clearsky.py

```diff
@@ -597,6 +599,264 @@ def _calc_d(aod700, p):
+    sky detection functions
+
+    Parameters
+    ----------
+    data : Series
+    samples_per_window : int
+        Number of data points in each window
+    sample_interval : float
+        Time in minutes in each sample interval
+    align : bool, default 'left'
+        Alignment of labels to data in sliding window. Must be one of
+        'left', 'center', 'right'.
+
+    Returns
+    -------
+    data_mean : Series
+        mean of data in each window
+    data_max : Series
+        maximum of data in each window
+    data_line_length : Series
+        length of line segments connecting successive data points
+    data_slope_nstd : Series
+        standard deviation of difference between data points in each window
+    data_slope : Series
+        difference between successive data points
+    """
+
+    shift = _shift_from_align(align, samples_per_window)
+    center = align == 'center'
+    data_mean = data.rolling(samples_per_window,
+                             center=center).mean().shift(shift)
+    data_max = data.rolling(samples_per_window,
+                            center=center).max().shift(shift)
+    # shift to get forward difference, .diff() is backward difference instead
+    data_diff = data.diff().shift(-1)
+    data_slope = data_diff / sample_interval
+    data_slope_nstd = data.rolling(samples_per_window, center=center).apply(
+        _calc_slope_nstd)
+    data_slope_nstd = data_slope_nstd.shift(shift)
+    data_line_length = data.rolling(samples_per_window, center=center).apply(
+        _calc_line_length, args=(sample_interval,))
+    data_line_length = data_line_length.shift(shift)
+
+    return (data_mean, data_max, data_line_length, data_slope_nstd, data_slope)
+
+
+def _calc_slope_nstd(data):
+    return np.diff(data).std(ddof=1) / data.mean()
+
+
+def _shift_from_align(align, window):
+    # number of places to shift to line up pandas.Series.rolling with
+    # align value.
+    # pandas.Series.rolling uses kwarg center with values of True or False
+    # default is False which means right aligned labels by default
+    # center=True aligns to center
+    # Here, calculate shift to align='left' (legacy pvlib behavior)
+    # code here for future capability if align='left', 'right' are of interest
+    # commented out to avoid decreasing test coverage
+    # if align == 'left':
+    #     shift = 1 - window
+    # else:
+    #     shift = 0
+    shift = 0
+    return shift
+
+
+def _count_in_window(ts, window):
+    # return int type instead of default int64, to avoid error in pandas
+    return int(ts.resample(window).agg('sum')[0])
+
+
+def _get_sample_intervals(times, window_length):
+    """ Calculates timeinterval and samples per window for Reno-style clear
+    sky detection functions
+    """
+    window = pd.Timedelta(minutes=window_length)
+    temp = pd.Series(data=1, index=times)
+
+    # determine if we can proceed
+    if times.inferred_freq and len(temp.unique()) == 1:
```

Comment:
This condition is opaque to me.  Won't `temp.unique()` always be `array([1])` if `times` is not empty?

Reply:
yes....a mistake